src/services/pcn-firewall/src/Chain.cpp

```diff
@@ -591,9 +591,156 @@ void Chain::delRuleList() {
 }
 
 ChainInsertOutputJsonObject Chain::insert(ChainInsertInputJsonObject input) {
-  throw std::runtime_error("Chain::ChainInsertOutput: Method not implemented");
+  ChainRuleJsonObject conf;
+
+  if (input.conntrackIsSet()) {
+    conf.setConntrack(input.getConntrack());
+  }
+  if (input.srcIsSet()) {
+    conf.setSrc(input.getSrc());
+  }
+  if (input.dstIsSet()) {
+    conf.setDst(input.getDst());
+  }
+  if (input.sportIsSet()) {
+    conf.setSport(input.getSport());
+  }
+  if (input.dportIsSet()) {
+    conf.setDport(input.getDport());
+  }
+  if (input.tcpflagsIsSet()) {
+    conf.setTcpflags(input.getTcpflags());
+  }
+  if (input.l4protoIsSet()) {
+    conf.setL4proto(input.getL4proto());
+  }
+  if (input.actionIsSet()) {
+    conf.setAction(input.getAction());
+  } else {
+    conf.setAction(ActionEnum::DROP);
+  }
+
+  uint32_t id = 0;
+  if (input.idIsSet()) {
+    id = input.getId();
+  }
+
+  if (id > rules_.size()) {
+    throw std::runtime_error("id not allowed");
+  }
+
+  auto newRule = std::make_shared<ChainRule>(*this, conf);
+
+  ChainStatsJsonObject confStats;
+  auto newStats = new ChainStats(*this, confStats);
```

Comment:
This should be wrapped in a unique_ptr because if one of the following calls throws an error this memory will be leaked.

Reply:
You are right, I wrapped it into a shared pointer for now, this should avoid possible memory leaks.